htlcswitch/circuit.go

```diff
@@ -53,82 +80,220 @@ func newPaymentCircuit(src, dest lnwire.ShortChannelID, key circuitKey,
+	return circuitKey(c.PaymentHash)
+}
+
+// circuitStorage is a data structure that implements thread safe storage of
 // circuits. Each circuit key (payment hash) may have several of circuits
 // corresponding to it due to the possibility of repeated payment hashes.
-//
-// TODO(andrew.shvv) make it persistent
-type circuitMap struct {
-	sync.RWMutex
-	circuits map[circuitKey]*paymentCircuit
+type circuitStorage struct {
+	db *channeldb.DB
 }
 
-// newCircuitMap creates a new instance of the circuitMap.
-func newCircuitMap() *circuitMap {
-	return &circuitMap{
-		circuits: make(map[circuitKey]*paymentCircuit),
-	}
+// newCircuitMap creates a new instance of the circuitStorage.
+func newCircuitMap(db *channeldb.DB) *circuitStorage {
+	return &circuitStorage{db}
 }
 
-// add adds a new active payment circuit to the circuitMap.
-func (m *circuitMap) add(circuit *paymentCircuit) error {
-	m.Lock()
-	defer m.Unlock()
-
-	// Examine the circuit map to see if this circuit is already in use or
-	// not. If so, then we'll simply increment the reference count.
-	// Otherwise, we'll create a new circuit from scratch.
-	//
-	// TODO(roasbeef): include dest+src+amt in key
-	if c, ok := m.circuits[circuit.PaymentHash]; ok {
-		c.RefCount++
-		return nil
-	}
+// add adds a new active payment circuit to the circuitStorage.
+func (m *circuitStorage) add(incomingCircuit *paymentCircuit) error {
+	return m.db.Batch(func(tx *bolt.Tx) error {
+		key := incomingCircuit.Key()
+		circuitBucket, err := tx.CreateBucketIfNotExists(circuitBucketKey)
+		if err != nil {
+			return err
+		}
+
+		counterBucket, err := tx.CreateBucketIfNotExists(counterCircuitBucketKey)
+		if err != nil {
+			return err
+		}
+
+		// Examine the circuit map to see if this circuit is already in use or
+		// not. If so, then we'll simply increment the reference count.
+		// Otherwise, we'll add new circuit in storage.
+		refCounter := uint64(1)
+		if data := counterBucket.Get(key[:]); data != nil {
+			refCounter = binary.BigEndian.Uint64(data)
+			refCounter++
+		}
+
+		// Update reference counter of the payment circuit.
+		var refCounterBytes [8]byte
+		binary.BigEndian.PutUint64(refCounterBytes[:], refCounter)
+		if err := counterBucket.Put(key[:], refCounterBytes[:]); err != nil {
+			return err
+		}
+
+		// If object have been putted in the storage already than exit,
+		// otherwise encode it and put it in the storage.
+		if refCounter != 1 {
+			return nil
+		}
 
-	m.circuits[circuit.PaymentHash] = circuit
+		// Encode the objects and place it in the circuitBucket.
+		var b bytes.Buffer
+		if err := incomingCircuit.Encode(&b); err != nil {
+			return err
+		}
 
-	return nil
+		return circuitBucket.Put(key[:], b.Bytes())
+	})
 }
 
-// remove destroys the target circuit by removing it from the circuit map.
-func (m *circuitMap) remove(key circuitKey) (*paymentCircuit, error) {
-	m.Lock()
-	defer m.Unlock()
+// remove destroys the target circuit by removing it from the circuit storage.
```

Comment:
looks like circuitKey doesn't get removed all the time right? ex. in the case `refCounter` isn't 0 on L263. Or should this never happen?